gentypes_test.go

```diff
@@ -0,0 +1,255 @@
+package gosysl
+
+import (
+	"go/format"
+	"testing"
+
+	"github.com/anz-bank/gosysl/pb"
+	testifyAssert "github.com/stretchr/testify/assert"
+)
+
+var primitiveTests = []struct {
+	input    pb.Type_Primitive
+	expected string
+}{
+	{pb.Type_BOOL, "bool"},
+	{pb.Type_ANY, "interface{}"},
+	{pb.Type_INT, "int"},
+	{pb.Type_STRING, "string"},
+	{pb.Type_FLOAT, "float64"},
+	{pb.Type_DECIMAL, "float64"},
+	{pb.Type_EMPTY, "nil"},
+	{pb.Type_BYTES, "[]byte"},
+	{pb.Type_DATE, "time.Time"},
+	{pb.Type_DATETIME, "time.Time"},
+	{pb.Type_STRING, "string"},
+}
+var primitiveErrTests = []pb.Type_Primitive{
+	pb.Type_XML,
+	pb.Type_UUID,
+}
+
+func TestGenPrimitiveType(tt *testing.T) {
+	assert := testifyAssert.New(tt)
+	for _, t := range primitiveTests {
+		str, err := GenPrimitiveType(&t.input)
+		assert.NoError(err)
+		assert.Equal(t.expected, str)
+	}
+	for _, tp := range primitiveErrTests {
+		_, err := GenPrimitiveType(&tp)
+		assert.Error(err)
+	}
+}
+
+func TestGenSimpleType(tt *testing.T) {
+	assert := testifyAssert.New(tt)
+	for _, t := range primitiveTests {
+		primitive := &pb.Type_Primitive_{Primitive: t.input}
+		pt := &pb.Type{Type: primitive}
+		str, err := GenSimpleType(pt)
+		assert.NoError(err)
+		assert.Equal(t.expected, str)
+	}
+	for _, tp := range primitiveErrTests {
+		primitive := &pb.Type_Primitive_{Primitive: tp}
+		pt := &pb.Type{Type: primitive}
+		_, err := GenSimpleType(pt)
+		assert.Error(err)
+	}
+
+	// test type refs
+	var refTests = []struct {
+		input    string
+		expected string
+	}{
+		{"MyType", "MyType"},
+		{"x", "x"},
+		{"map of string:int", "map[string]int"},
+	}
+	for _, t := range refTests {
+		ref := &pb.Scope{Path: []string{t.input}}
+		typeRef := &pb.Type_TypeRef{TypeRef: &pb.ScopedRef{Ref: ref}}
+		pt := &pb.Type{Type: typeRef}
+		str, err := GenSimpleType(pt)
+		assert.NoError(err)
+		assert.Equal(t.expected, str)
+	}
+	var errRefTests = []struct {
```

Comment:
Could use `[][]string` (It's a toss up whether consistency trumps simplicity, though.)

Reply:
I like that! Done.